src/network/webtask.cpp

```diff
@@ -310,12 +309,20 @@ void WebTask::slotNetworkReplyFinished() {
         DEBUG_ASSERT(m_timeoutTimerId == kInvalidTimerId);
         kLogger.debug()
                 << this
-                << "Ignoring obsolete network reply";
+                << "Discarding obsolete network reply";
+        finishedNetworkReply->deleteLater();
         return;
     }
     VERIFY_OR_DEBUG_ASSERT(m_pendingNetworkReply == finishedNetworkReply) {
+        kLogger.debug()
+                << this
+                << "Discarding unexpected network reply";
+        finishedNetworkReply->deleteLater();
         return;
     }
+    // As a side-effect QObject::deleteLater() might reset
+    // QPointer references like m_pendingNetworkReply!
+    finishedNetworkReply->deleteLater();
     m_pendingNetworkReply = nullptr;
```

Comment:
I'm a bit confused. Here you are explicitly setting `m_pendingNetworkReply` to nullptr, so isn't it always reset?

Reply:
Just to ensure that it is not used after invoking deleteLater(). `finishedNetworkReply` is valid until the end of the scope while `m_pendingNetworkReply` might become a `nullptr` earlier!

Reply:
Shouldn't `m_pendingNetworkReply` be set to nullptr before the earlier return statements then?

Reply:
When it becomes a nullptr depends on how QNetworkManager is implemented and how long the actual reference is alive. We don't want such an uncontrollable dependency that could easily break at any time.

I have experienced this surprising (but valid) behavior of QPointer in a similar situation. Explicitly resetting it prevents uncontrollable side-effects depending on the circumstances.

Reply:
Are we in finishedNetworkReply->thread()? If not all accesses to finishedNetworkReply are on a risk below. 
I think it is better to move this to the very end in any case.

Reply:
m_pendingNetworkReply = nullptr; should be done before the delete later call. 

I am not sure about the threading, but things like 
if (m_pendingNetworkReply) {
    m_pendingNetworkReply->doSomthing()
}

Are only save in m_pendingNetworkReply->thread(), because else the deleteLater() delete can be executed in between.

Reply:
OT: it would be good if we document what thread a method is executed in inside a doxygen comment.

Reply:
I disagree. Only exceptional behavior should be documented. A QObject with slots is only supposed to be accessed from the event loop thread by convention.

Reply:
@daschuer This is a slot handler, so we are in the event loop thread.

Reply:
We are supposed to access the sender QNetworkReply here in this slot and don't need to care which thread it actually lives in. It is managed by Qt. The pointer obtained by `sender()` won't disappear until the method is finished. But this assumption is wrong for the QPointer after invoking deleteLater(), therefore we invalidate it manually!!

Sorry, I didn't design this confusing and dangerous behavior. That's how Qt works.

Reply:
The order of deletion doesn't matter here. WebTask has been designed for Qt's event loop paradigm and is not accessed by any other thread concurrently. Different from a lot of Mixxx code that violates the basic rule of not sharing those objects between threads except for very rare cases that require and exception and need special care.

Reply:
> Shouldn't `m_pendingNetworkReply` be set to nullptr before the earlier return statements then?



> Shouldn't `m_pendingNetworkReply` be set to nullptr before the earlier return statements then?

No. Only if it equals the sender() object that gets deleted by deleteLater().

Reply:
> We are supposed to access the sender QNetworkReply here in this slot and don't need to care which thread it actually lives in. 

I am afraid this is not entirely true. This slot is called in the WebTask thread. The delete later is executed in the QNetworkReply thread. If the deleteLater() performs the delete before this lot here is over, m_pendingNetworkReply becomes null and finishedNetworkReply becomes a dangling pointer. The next access will crash Mixxx. 

If both objects live in teh same thread this cannot happen. I am not able to see instantly if this is the case.
In any case it is no harm to change the execution odrder as proposed.  

> It is managed by Qt. The pointer obtained by sender() won't disappear until the method is finished. 

Where do you take this info from? 

> But this assumption is wrong for the QPointer after invoking deleteLater(), therefore we invalidate it manually!!

As Far es I know the QPointer is set to null in the destructor of the object. I don't know that there is a lock or anything waiting for pending signals. Is is it? 

But anyway lets change the execution order and not rely on Qts implementation details.

Reply:
Not true. The slot is called from the event loop and deleteLater() will not delete the object until control returns back to the event loop. But *weak pointers* like QPointer might be invalidated immediately, even if the object itself is still alive.

Reply:
This is all documented in QObject::deleteLater().

Reply:
Qt is allowed to invalidate weak pointers after QObject::deleteLater() has been called to prevent handing out new plain pointers that are expected to become invalid soon. Perfectly reasonable, and I have experienced this behavior already.

Reply:
Please clarify: Do the WebTask and m_pendingNetworkReply live in the same thread?

Reply:
Yes. WebTask is a QObject and all methods are accessed **only** by the event loop thread. 

Only the public invokeStart()/invokeAbort() are allowed to be invoked from different threads, because they delegate to QMetaObject::invokeMethod().

Reply:
Why do I need to clarify the decision to use Qt as it is supposed to be used, i.e. signal-threaded with pure signal/slot messaging?

Reply:
First of all, since sender and receiver live in the same thread here, we have no issue. 
There should be no point in this whole PR. 

> After invoking QObject::deleteLater() on an object all weak QPointer references may become nullptr, either immediately or later.

This is only true if the sender and receiver live in different threads. This is not the case here. 

The delete later implementation is quite simple. It calls: 

```
void QObject::deleteLater()
{
    QCoreApplication::postEvent(this, new QDeferredDeleteEvent());
}

deleteLaterCalled
```

Reply:
If you want to make the implementation save in a multi threads context. You need to move deleteLater() to the very end. 
If not, we can IMHO close this PR. 

But probably I miss something. Do I?